### PR TITLE
[SPARK-33771][SQL][TESTS] Fix Invalid value for HourOfAmPm when testing on JDK 14

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr fix invalid value for HourOfAmPm when testing on JDK 14.

### Why are the changes needed?

Run test on JDK 14.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

N/A

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/TimestampFormatterSuite.scala [hunk 1]
```diff
@@ -355,9 +356,14 @@ class TimestampFormatterSuite extends DatetimeFormatterSuite {
...
+      // JDK-8223773: DateTimeFormatter Fails to throw an Exception on Invalid HOUR_OF_AMPM
       // For `KK`, "12:00:00 am" is the same as "00:00:00 pm".
-      val micros2 = formatter.parse("2009-12-12 12 am")
-      assert(micros2 === date(2009, 12, 12, 12))
+      if (SystemUtils.isJavaVersionAtLeast(JavaVersion.JAVA_13)) {
+        intercept[DateTimeException](formatter.parse("2009-12-12 12 am"))
+      } else {
+        val micros2 = formatter.parse("2009-12-12 12 am")
+        assert(micros2 === date(2009, 12, 12, 12))
+      }
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/TimestampFormatterSuite.scala:366]: This is a JDK bug, fixed by https://bugs.openjdk.java.net/browse/JDK-8223773

**Review by dongjoon-hyun — COMMENTED** [review]
Thanks, @wangyum . Is this the only failure at JDK14+?

**Comment by HyukjinKwon** [comment]
Do we plan to support JDK 14? BTW I remember @srowen mentioned few more tests:

```
- ZooKeeperPersistenceEngine *** FAILED ***
  org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss for /spark/master_status

- parsing hour with various patterns *** FAILED ***
  java.time.format.DateTimeParseException: Text '2009-12-12 12 am' could not be parsed: Invalid value for HourOfAmPm (valid values 0 - 11): 12
```

before. Let's fix all or none. We don't even know if we're going to support JDK 14.

**Comment by dongjoon-hyun** [comment]
We will not support JDK14 officially. It's just one of EOL releases. I believe this is an effort to reduce the gap between `Java 11` and `Java 17 (LTS) September 2021`.

**Comment by dongjoon-hyun** [comment]
Let me make it official by filing an official JIRA.

**Comment by srowen** [comment]
I mean, we should definitely support future Java versions. Fortunately, there seems to be very few problems when testing vs Java 14, because Java is not changing nearly as much after Java 9 as it did between 8 and 9. Yes this was one I observed too. I do think we should target support for LTS Java releases like Java 14. If this one is just due to a JDK bug, fine, but we should be testing vs Java 14 now and vs Java 17 upon its release.

**Comment by dongjoon-hyun** [comment]
Ur, Java 14 is not LTS, isn't it?
>  we should target support for LTS Java releases like Java 14.

**Comment by dongjoon-hyun** [comment]
I filed a JIRA, @srowen , @wangyum , @HyukjinKwon . We can collect all our effort and discussion there.
- https://issues.apache.org/jira/browse/SPARK-33772

**Comment by srowen** [comment]
Oh OK, 14 is not an LTS release? I thought every 3rd major release was from here. Well it won't matter much. I think supporting 17 is fine and will require few if any changes.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/TimestampFormatterSuite.scala (hunk 1 above)
Comment (srowen) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/TimestampFormatterSuite.scala:366]: I think this is OK.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I'm good with this change too.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum , @srowen , @HyukjinKwon .

**Pull request closed without merging**